Methods for extracting boundary edge DOFs from a 3D ParMesh and for synchronizing boundary conditions across processor boundaries

dnpham23 opened the pull request:
This PR introduces new methods for implementing boundary conditions on 1D loops in a 3D ParMesh. The key additions are:
- Method `GetBoundaryEdgeDoFs` and supporting methods added to `pfespace` to extract the edge DoFs on the 1D perimeters of 2D surfaces embedded in 3D mesh. 
- Reduce operation `MaxAbs` that performs maximum absolute value reduction operation for MPI communication. This is needed to allow proper identification of boundary values assigned to the ldofs that are shared by multiple processors. 
- Selective MPI reduction operation `ReduceMarked` (a variation of `ReduceEnd`) that only applies the reduction operation to DOFs marked in a boolean array. This helps avoid unnecessary communications in cases where only a small number of DoFs need to be communicated across ranks. 
- Unit tests for edge dof extractions and reduction operation.
- A mesh `nested_cubes` that allows for convenient checking of internal boundaries.

### Usage of this feature
In some of the EM applications it is desirable to apply boundary conditions on 1D curves and loops that are parts of the boundaries on 2D surfaces embedded in 3D structures. an example could be boundary conditions on the inner perimeter of a 2D disk containing a hole:
<img width="1840" height="1620" alt="circular_hole" src="https://github.com/user-attachments/assets/ef9f8f5a-3545-4bc4-96e9-a46672b9eacc" />

In such cases it is usually more convenient to apply boundary conditions on the ldofs and the transfer them over to tdof. This requires proper synchronization to handle edge cases where the dofs are shared by multiple ranks to avoid lost of BC information. The illustration below shows a situation when that can happen, using a structure containing nested cubes.

The example geometry:
<img width="1148" height="1080" alt="image (4)" src="https://github.com/user-attachments/assets/06b887ef-b618-4488-b50c-7ef36f8cec7b" />

Without correct reduction and synchronization:
<img width="800" height="700" alt="intbdr_xz_p_numproc6_2ndorder_nosync" src="https://github.com/user-attachments/assets/0ec8b111-3805-4768-ae36-98509e3f2169" />

With correct reduction and synchronization:
<img width="800" height="700" alt="intbdr_xz_p_numproc6_2ndorder" src="https://github.com/user-attachments/assets/f77be2e8-9fb9-4531-b81c-ec65a3abf003" />

<!--GHEX{"author":"dnpham23","assignment":"2026-07-28T21:52:37","editor":"tzanio","id":4983,"reviewers":["psocratis","dylan-copeland","tzanio","mlstowell","najlkin","hughcars"],"merge":"2026-08-18T21:52:37","approval":"2026-08-11T21:52:37"}XEHG-->
<!--GHEXTABLE-->
 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#4983](https://github.com/mfem/mfem/pull/4983) | @dnpham23 | @tzanio | @psocratis + @dylan-copeland + @tzanio + @mlstowell + @najlkin + @hughcars | 7/28/26 | ⌛due 8/11/26 | ⌛due 8/18/26 | |
<!--ELBATXEHG-->
<details>
<summary>PR Checklist</summary>
    
- [ ] Code builds.
- [ ] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [ ] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [ ] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [ ] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [ ] Run `make unittest` to make sure all unit tests pass.
- [ ] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

dnpham23 committed:
Add methods to extract boundary edge dofs and loop orientations

dnpham23 committed:
Add brute force unit tests

dnpham23 committed:
Add patch files to include edge dof methods into pfespace

dnpham23 committed:
Add new mesh for testing

dnpham23 committed:
Add new test for loop edge boundary condition

dnpham23 committed:
Add method to compute perimeters of bounary loops

dnpham23 committed:
Update Makefile for new test, modify cube test to use edge dof extraction from mfem

dnpham23 committed:
Add new synchronize method for applying loop BC on tdofs in parallel. Add new MaxAbs reduce operation

dnpham23 committed:
Clean up the cube test

dnpham23 committed:
Modify GetBoundaryEdgeDoFs to return ldof markers, add new reduction method ReduceMarked

dnpham23 committed:
Update unit tests and cubetest to use new GetBoundaryEdgeDoFs and ReduceMarked (if applicable)

dnpham23 committed:
Add alternative method to synchronize only marked ldofs across processors.

dnpham23 committed:
WIP

dnpham23 committed:
generalize the edge dof extraction method to multiple boundary loops

dnpham23 committed:
update cubetest

dnpham23 committed:
Update tests for boundary edge DoFs

dnpham23 committed:
remove out-of-date manual patch files

dnpham23 committed:
clean up and reorganize the tests for boundary edge dof methods

dnpham23 committed:
add unit tests for edge dof extractions and synchronization

dnpham23 committed:
remove temporary test folder

dnpham23 committed:
fix format, add missing documentation, fix variable shadow

hughcars commented on fem/pfespace.cpp:
```diff
@@ -1255,6 +1263,372 @@ void ParFiniteElementSpace::GetExteriorVDofs(Array<int> &ext_dofs,
+
+   // Get global indices
+   Array<HYPRE_BigInt> global_edge_indices;
+   pmesh->GetGlobalEdgeIndices(global_edge_indices);
```
Add an assertion at the start of this method that `!pmesh->Nonconforming()`, as these calls will ultimately fail because of that.

hughcars commented on fem/pfespace.cpp:
```diff
@@ -1255,6 +1263,372 @@ void ParFiniteElementSpace::GetExteriorVDofs(Array<int> &ext_dofs,
+                                                std::unordered_map<int, int> *dof_to_boundary_element_out,
+                                                Array<int> *ess_edge_list)
+{
+   // Data structures for boundary edge DoFs
```
This is concerned with Dimension 3 meshes finding Dimension 1 objects, so add some verify calls that the underlying mesh is dim 3 at least. 

Additionally, the initial serial piece of this should be made into a method within `FiniteElementSpace` that is called from here. So if `GetBoundaryEdgeDoFs` is virtual, the serial version contains the first piece of this, then the parallel version here calls `FiniteElementSpace::GetBoundaryEdgeDoFs` then does the parallel discovery portion here.

hughcars commented on fem/pfespace.cpp:
```diff
@@ -1255,6 +1263,372 @@ void ParFiniteElementSpace::GetExteriorVDofs(Array<int> &ext_dofs,
+
+   // Pre-compute face indices for boundary elements
+   std::unordered_map<int, int> boundary_element_to_face;
+   for (int i = 0; i < boundary_element_indices.Size(); ++i)
+   {
+      int boundary_element_idx = boundary_element_indices[i];
```
nit: use range loops `for (int boundary_element_idx : boundary_element_indices)`

hughcars commented on fem/pfespace.cpp:
```diff
@@ -1255,6 +1263,372 @@ void ParFiniteElementSpace::GetExteriorVDofs(Array<int> &ext_dofs,
+      }
+
+      // Process collected data
+      std::unordered_map<HYPRE_BigInt, std::unordered_set<HYPRE_BigInt>>
+                                                                      edge_to_faces;
```
strange indenting?

dnpham23 replied:
don't know what happened there

hughcars commented on fem/pfespace.cpp:
```diff
@@ -1255,6 +1263,372 @@ void ParFiniteElementSpace::GetExteriorVDofs(Array<int> &ext_dofs,
+         edge_to_faces[all_data[i]].insert(all_data[i + 1]);
+      }
+
+      // Mark DoFs from artificial edges for removal
```
What is an artificial edge here? Clarify in the comment

hughcars commented on fem/pfespace.cpp:
```diff
@@ -1255,6 +1263,372 @@ void ParFiniteElementSpace::GetExteriorVDofs(Array<int> &ext_dofs,
    Synchronize(ext_dofs);
 }
 
+void ParFiniteElementSpace::GetBoundaryEdgeDoFs(const Array<int>
```
Nit: `Dofs` not `DoFs`

hughcars commented on fem/pfespace.cpp:
```diff
@@ -1255,6 +1263,372 @@ void ParFiniteElementSpace::GetExteriorVDofs(Array<int> &ext_dofs,
+   std::unordered_set<int> processed_edges;
+   processed_edges.reserve(boundary_edge_dofs.size());
+
+   for (const auto& pair : dof_to_edge_map)
+   {
+      int local_edge = pair.second;
```
mfem is using c++17 now, so I think you can do `const auto &[dof, local_edge] : dof_to_edge_map)` to bind the variables directly.

hughcars commented on fem/pfespace.cpp:
```diff
@@ -1255,6 +1263,372 @@ void ParFiniteElementSpace::GetExteriorVDofs(Array<int> &ext_dofs,
+   std::unordered_set<int> boundary_edge_dofs;
+   std::unordered_map<int, int> dof_to_edge_map;
+   std::unordered_map<int, int> dof_to_boundary_element;
+   std::unordered_map<int, int> dof_to_edge_orientation;
```
`dof_to_edge_map` can be used to encode the orientation using the standard mfem trick, if orientation is positive, then use the index, if it's negative then use `-1-i`. That way only one map is needed.

najlkin replied:
+1 from here, `dof_to_(edge_)orientation` can be completely eliminated

dnpham23 replied:
Done! `dof_to_orientation` is gone. Orientation is computed on demand by `ComputeLoopEdgeOrientations`, which returns a parallel-indexed Array.

hughcars commented on fem/pfespace.cpp:
```diff
@@ -1255,6 +1263,372 @@ void ParFiniteElementSpace::GetExteriorVDofs(Array<int> &ext_dofs,
+   {
+      ldof_marker[dof] = 1; // Mark all boundary edge dofs
+      int tdof = GetLocalTDofNumber(dof);
+      if (tdof >= 0)
```
Add a trailing comment noting that `tdof == -1` means not present on this rank.

hughcars commented on fem/pfespace.cpp:
```diff
@@ -1255,6 +1263,372 @@ void ParFiniteElementSpace::GetExteriorVDofs(Array<int> &ext_dofs,
+   std::unordered_map<int, int>& edge_loop_orientations)
+{
+   // Process each edge locally
+   for (const auto& pair : dof_to_boundary_element)
```
Can use structured binding `for (const auto &[dof, bdr_elem_idx] : dof_to_boundary_element)`

hughcars commented on fem/pfespace.cpp:
```diff
@@ -1255,6 +1263,372 @@ void ParFiniteElementSpace::GetExteriorVDofs(Array<int> &ext_dofs,
+      int edge_id = edge_it->second;
+
+      // Get edge vertices
+      Array<int> edge_verts;
```
Hoist these `Array<int>` outside of the loop to avoid repeated heap allocations.

hughcars commented on fem/pfespace.cpp:
```diff
@@ -1255,6 +1263,372 @@ void ParFiniteElementSpace::GetExteriorVDofs(Array<int> &ext_dofs,
+         }
+      }
+
+      if (third_vertex == -1) { continue; } // No third vertex found, skip this edge
```
This seems like it should be a hard error rather than a skip?

hughcars commented on fem/pfespace.cpp:
```diff
@@ -1255,6 +1263,372 @@ void ParFiniteElementSpace::GetExteriorVDofs(Array<int> &ext_dofs,
+      const double *v2 = pmesh->GetVertex(third_vertex);
+
+      // Edge vector
+      Vector edge_vec(3);
```
Lift these two `Vector` outside the loop, for same reason as `Array<int>`

hughcars commented on fem/pfespace.hpp:
```diff
@@ -448,6 +454,49 @@ class ParFiniteElementSpace : public FiniteElementSpace
+       @param[in] boundary_element_indices Array of boundary element indices
+       @param[out] ess_tdof_list Array of essential true DOF indices
+       @param[out] ldof_marker Array marking which local DOFs are boundary edge DOFs
+       @param[out] boundary_edge_dofs_out set of boundary edge DOFs
+       @param[out] dof_to_edge Optional map from DOFs to edge indices
+       @param[out] dof_to_orientation Optional map from DOFs to edge orientations
+       @param[out] dof_to_boundary_element_out Optional map from DOFs to boundary elements
+       @param[out] ess_edge_list Optional array of edge indices */
+   void GetBoundaryEdgeDoFs(const Array<int> &boundary_element_indices,
+                            Array<int> &ess_tdof_list,
+                            Array<int> &ldof_marker,
+                            std::unordered_set<int> &boundary_edge_dofs_out,
+                            std::unordered_map<int, int> *dof_to_edge = nullptr,
+                            std::unordered_map<int, int> *dof_to_orientation = nullptr,
+                            std::unordered_map<int, int> *dof_to_boundary_element_out = nullptr,
+                            Array<int> *ess_edge_list = nullptr);
+
+   /** Find the boundary elements marked with specified boundary attributes
+        @param[in] bdr_attrs list of boundary attributes to search for
+        @param[out] attr_to_elements map from boundary attribute to the list of
+                                     boundary elements with that attribute */
+   void GetBoundaryElementsByAttribute(const Array<int> &bdr_attrs,
+                                       std::unordered_map<int, Array<int>> &attr_to_elements);
+
+   /// @brief Simplified interface for single boundary attribute
+   void GetBoundaryElementsByAttribute(int bdr_attr,
+                                       Array<int> &boundary_elements);
+
+   /** Compute edge orientations relative to a loop direction defined by a normal vector.
+       This is useful for applying circulation boundary conditions.
+       @param[in] dof_to_edge Map from DOFs to edge indices
+       @param[in] dof_to_boundary_element Map from DOFs to boundary elements
+       @param[in] loop_normal Normal vector defining the loop direction
+       @param[out] edge_loop_orientations Map from edge indices to orientations (+1 or -1) */
+   void ComputeLoopEdgeOrientations(const std::unordered_map<int, int>&
+                                    dof_to_edge,
+                                    const std::unordered_map<int, int>& dof_to_boundary_element,
+                                    const Vector& loop_normal,
+                                    std::unordered_map<int, int>& edge_loop_orientations);
+
```
These all probably need `FiniteElementSpace` counterparts, which are virtual methods that execute the serial portions, then in `ParFiniteElementSpace` there would be an `override` that first dispatches to the `FiniteElementSpace` version before completing to parallelism.

hughcars commented on tests/unit/fem/test_bdr_edgedof.cpp:
```diff
@@ -0,0 +1,337 @@
+{
+   std::vector<int> partition(n_elements);
+
+   std::function<void(int)> generate = [&](int elem)
```
Generally you just write `auto generate = [&]` to define a lambda, the type is then unique to the call site and the compiler does the hard work.

hughcars commented on tests/unit/fem/test_bdr_edgedof.cpp:
```diff
@@ -0,0 +1,337 @@
+} // namespace boundary_edge_dof_test
+
+TEST_CASE("BoundaryEdgeDoFsPartitionInvariant",
+          "[Parallel], [ParMesh], [BoundaryEdgeDoFs]")
```
The second arg here is a different format to below, which is correct?

dnpham23 replied:
I think so? fixed the format inconsistency anyway

hughcars reviewed (changes requested):
An initial pass, generally LGTM. Would be nice to attach some images of the edge dof selection to the pr too, to help visualize some usage of this feature.

dnpham23 committed:
WIP - refactoring

dnpham23 committed:
Refactoring

dylan-copeland commented on tests/unit/fem/test_bdr_edgedof.cpp:
```diff
@@ -0,0 +1,337 @@
+TEST_CASE("BoundaryEdgeDoFsPartitionInvariant",
+          "[Parallel][ParMesh][BoundaryEdgeDoFs]")
+{
+   const int orientation = 3;
```
`constexpr`

dylan-copeland commented on tests/unit/fem/test_bdr_edgedof.cpp:
```diff
@@ -0,0 +1,337 @@
+}
+
+// Helper function to compute boundary loop length
+double ComputeBoundaryLoopLength(ParMesh* pmesh,
```
`double` -> `real_t`

dylan-copeland commented on fem/pfespace.hpp:
```diff
@@ -420,6 +422,10 @@ class ParFiniteElementSpace : public FiniteElementSpace
 
+   /** @brief Given a vector of boundary condition values on the local true
+     dofs, perform a reduction across all processors. */
+   void SynchronizeBC(Array<double> &bc_values) const;
```
`real_t` should be used instead of `double` everywhere in MFEM.

dylan-copeland commented on tests/unit/fem/test_bdr_edgedof.cpp:
```diff
@@ -0,0 +1,337 @@
+namespace boundary_edge_dof_test
+{
+
+// Generate all possible partitionings of n_elements into num_procs
```
The current MFEM unit tests run quickly, and we should try to keep it that way. These unit tests are too slow in parallel, as the number of cases produced by using all possible partitions is extremely large, even for 2 MPI ranks. For example, 

srun -n 2 punit_tests "[BoundaryEdgeDoFs]"
All tests passed (131170 assertions in 3 test cases)

Can this test be reduced to a small number of cases that test the capability sufficiently?

dylan-copeland commented on data/nested_cubes.msh:
```diff
@@ -0,0 +1,2300 @@
+$MeshFormat
```
It seems this mesh is added, because the inner cube is difficult to generate and cannot be done simply with code such as `MakeCartesian3D`. If it is necessary to add such a mesh, is it possible to make a coarser version, to keep the file added to the repo of minimal size? If the tests should use a finer mesh, it would be preferable to generate the coarsest mesh possible and refine it within the code.

dylan-copeland commented on fem/pfespace.hpp:
```diff
@@ -448,6 +454,49 @@ class ParFiniteElementSpace : public FiniteElementSpace
+                            std::unordered_map<int, int> *dof_to_edge = nullptr,
+                            std::unordered_map<int, int> *dof_to_orientation = nullptr,
+                            std::unordered_map<int, int> *dof_to_boundary_element_out = nullptr,
+                            Array<int> *ess_edge_list = nullptr);
```
`override`?

dylan-copeland commented on fem/fespace.hpp:
```diff
@@ -1400,6 +1401,33 @@ class FiniteElementSpace
+   virtual void GetBoundaryElementsByAttribute(int bdr_attr,
+                                               Array<int> &boundary_elements);
+
+   /** @brief Compute edge orientations for loop traversal.
```
It would be helpful to add an explanation of what `loop` means in this context.

dylan-copeland commented on fem/pfespace.cpp:
```diff
@@ -1118,6 +1118,13 @@ void ParFiniteElementSpace::Synchronize(Array<int> &ldof_marker) const
    gcomm->Bcast(ldof_marker);
 }
 
+void ParFiniteElementSpace::SynchronizeBC(Array<double> &bc_values) const
```
Where is this used?

dylan-copeland commented on fem/fespace.cpp:
```diff
@@ -4497,6 +4499,160 @@ ::GetEssentialBdrEdgesFaces(const Array<int> &bdr_attr_is_ess,
+            }
+            else
+            {
+               // DoF appears twice - interior to boundary, remove it
```
It appears this function only finds edges on the boundary of the boundary. For example, if there is a square cut on the interior of a 3D domain, this function will find edges on the 4 sides of the interior square. If the interior boundary is the surface of a cube inside the domain (like the plot shown in the PR description), it will not find any edges. Is this right? In any case, more documentation for this function in the header file would be helpful, to explain exactly what it does.

dnpham23 committed:
update unit test to reduce # of cases, replace test mesh with a leaner one, add documentation and some minor changes

dnpham23 committed:
extend GetBoundaryEdgeDofs to cover both 2D and 3D cases, rename it to GetBoundaryLoopEdgeDofs, remove SynchronizeBC, add unit test for 2D case

dylan-copeland commented on fem/pfespace.cpp:
```diff
@@ -1255,6 +1255,361 @@ void ParFiniteElementSpace::GetExteriorVDofs(Array<int> &ext_dofs,
+   const Array<int> &bdr_attrs,
+   std::unordered_map<int, Array<int>> &attr_to_elements)
+{
+   // Initialize arrays for each attribute
```
The implementation of this function is the same as for `FiniteElementSpace::GetBoundaryElementsByAttribute`, except for the usage of `mesh` or `pmesh`. The functions called for the mesh is in the base `class Mesh`. Would it be simpler to put this function in `Mesh`? Alternatively, a function could be defined, not in a class, taking a `Mesh` argument, and called by `GetBoundaryElementsByAttribute` in `FiniteElementSpace` and `ParFiniteElementSpace` with `mesh` or `pmesh`.

dnpham23 replied:
Thanks for pointing out. I followed the suggested alternative approach, defining a helper function taht takes a Mesh argument and called by GetBoundaryElementsByAttribute.

dylan-copeland commented on fem/pfespace.cpp:
```diff
@@ -1255,6 +1255,361 @@ void ParFiniteElementSpace::GetExteriorVDofs(Array<int> &ext_dofs,
+   const Vector& loop_normal,
+   std::unordered_map<int, int>& edge_loop_orientations)
+{
+   Array<int> edge_verts, bdr_elem_verts;
```
This function also appears to have the same implementation as for `FiniteElementSpace`, with `pmesh` instead of `mesh`. Can it be refactored to avoid duplication?

dylan-copeland commented on general/communication.hpp:
```diff
@@ -416,6 +416,43 @@ class GroupCommunicator
    template <class T> static void Max(OpData<T>);
    /// Reduce operation bitwise OR, instantiated for int only
    template <class T> static void BitOR(OpData<T>);
+   /// Reduce operation max absolute value, instantiated for int and double
```
Since `SynchronizeBC` has been removed, it seems that the additions to `GroupCommunicator` could also be removed, but maybe it should be kept anyway. If it could be useful in the future, we could keep it instead of throwing away something useful. One question I have about the usage is whether this functionality could be accomplished easily by just taking the absolute value of an array and then using the usual max operator. This approach has the disadvantage of requiring a copy of the array (which may be enough to justify `MaxAbs`), but is there another reason to use the new `MaxAbs` operator?

dnpham23 replied:
`MaxAbs` operation provides a rather unique functionality that could be useful in various situations. A few scenarios where it is convenient to have `MaxAbs` includes:
1. Boundary conditions implementation: there are situations where, for example, signed boundary values imposed on ldofs need to be inherited by tdofs, then MaxAbs returns both the magnitude and the correct sign of the boundary value.
2. Error estimations: MaxAbs preserves whether the largest error is positive or negative. For example: 
	double max_error = /* MaxAbs across processors */;
	if (max_error < 0) {
	 // Largest error indicates underestimation
	} 
	else {
	// Largest error indicates overestimation  
	}
3. Stability monitoring: in iterative solvers, the sign indicates whether the instability is growing in positive or negative direction.
Workarounds do exist, but are less efficient. They require, as you have pointed out, a copy of the array otherwise the information on the signs are lost. Otherwise you’d also probably need two reductions instead of just one. For example, given a dataset like:
rank 0: [-5.0, 2.0]
rank 1: [3.0, -1.0] 
rank 2: [1.5, -4.0]
MaxAbs returns -5.0, and to achieve the same result you either have to (1) separate positive and negative reductions then compare the magnitudes of the two or (2) do the Max reduction and Min reduction and then compare the two.

dylan-copeland replied:
It would be helpful to mention in the comment that the result is signed.

dylan-copeland replied:
Also `float` can be mentioned.

dnpham23 replied:
Good point. Sum/Min/Max/MaxAbs are all instantiated for float as well. Updated the four comments to say "int, double and float" (BitOR stays int-only). I also added a `@note` that the reduction is carried out in the signed type T, so the result is signed even for bitwise operations.

dnpham23 commented on fem/pfespace.cpp:
```diff
@@ -1255,6 +1255,361 @@ void ParFiniteElementSpace::GetExteriorVDofs(Array<int> &ext_dofs,
+                                                    std::unordered_map<int, int> *dof_to_boundary_element_out,
+                                                    Array<int> *ess_edge_list)
+{
+   MFEM_ASSERT(!pmesh->Nonconforming(),
```
updated

dnpham23 commented on fem/pfespace.cpp:
```diff
@@ -1255,6 +1255,361 @@ void ParFiniteElementSpace::GetExteriorVDofs(Array<int> &ext_dofs,
+            if (edge_to_faces[global_edge_id].size() >= 2)
+            {
+               auto it = global_to_local_edge.find(global_edge_id);
+               if (it != global_to_local_edge.end())
```
Thanks for pointing out, the check was just defensive programming that wasn't needed. The minimal fix removes the iterator check and directly uses the map lookup with [], since we know the key exists and all entries in `local_data` came from `global_edge_indices` which is used to populate `global_to_local_edge`.

dnpham23 committed:
refactoring functions in fespace and pfespace

psocratis commented on tests/unit/fem/test_bdr_edgedof.cpp:
```diff
@@ -0,0 +1,465 @@
+   serial_mesh.UniformRefinement();
+   
+   int num_procs = Mpi::WorldSize();
+   int rank = Mpi::WorldRank();
```
`rank` is not used

psocratis commented on tests/unit/fem/test_bdr_edgedof.cpp:
```diff
@@ -0,0 +1,465 @@
+      all_dof_results.reserve(all_partitionings.size());
+      
+      // Test each partitioning
+      int partition_idx = 0;
```
unused variable

dnpham23 committed:
Merge branch 'master' into master

dnpham23 committed:
Apply code style formatting to boundary edge DoF files and tests

dnpham23 committed:
clean up unused variables

dnpham23 committed:
Merge remote-tracking branch 'upstream/master'

hughcars committed:
Merge branch 'master' into master

hughcars commented on general/communication.hpp:
```diff
@@ -416,6 +416,43 @@ class GroupCommunicator
+         case Sum_Op:    Op = GroupCommunicator::Sum; break;
+         case Min_Op:    Op = GroupCommunicator::Min; break;
+         case Max_Op:    Op = GroupCommunicator::Max; break;
+         case BitOR_Op:  Op = GroupCommunicator::BitOR; break;
```
Instantiating this overload for `double` or `float` still forms a reference to `BitOR<T>`, even when another operation is selected. Since only `BitOR<int>` is instantiated, an unoptimized floating-point caller fails to link. Discard that branch at compile time for non-integral types.

```suggestion
         case BitOR_Op:
            if constexpr (std::is_integral<T>::value)
            {
               Op = GroupCommunicator::BitOR;
            }
            else
            {
               MFEM_ABORT("BitOR reduction requires an integral type.");
            }
            break;
```

dnpham23 replied:
Good catch, fixed in [dae8a18](https://github.com/mfem/mfem/pull/4983/commits/dae8a18c323d4adab99c150a0c18ff454a93f01d) using your suggested `if constexpr` guard.

hughcars commented on fem/pfespace.cpp:
```diff
@@ -1285,6 +1285,280 @@ void ParFiniteElementSpace::GetExteriorVDofs(Array<int> &ext_dofs,
+      {
+         int edge = dof_to_edge_map[dof];
+         tdof_edge_pairs.push_back({tdof, edge});
+      }
```
How is a selected shared DoF emitted when its boundary face is on a non-master rank? `GetLocalTDofNumber` returns `-1` there, while the master may own no selected face, so neither rank appends it. Is a synchronization step missing before converting the local marker to true DoFs? The current tests do not inspect `ess_tdof_list`.

dnpham23 replied:
Confirmed, a synchronization step was indeed missing. Fixed in [dae8a18](https://github.com/mfem/mfem/pull/4983/commits/dae8a18c323d4adab99c150a0c18ff454a93f01d)
(pfespace.cpp:1523).

You were also right that the existing
tests don't inspect `ess_tdof_list`. I wrote a probe that compares the set of
selected shared DoFs against the set actually emitted, identifying DoFs by
`GetGlobalTDofNumber` so the comparison is rank-agnostic. On 3D
`OrientedTriFaceMesh` across three partitionings it found 2 DoFs (4 on five
ranks) that were emitted by no rank — precisely the case you described.

The fix marks the local DoFs, then calls `Synchronize(ldof_marker)` before
converting to true DoFs, and iterates the marker rather than the local set.

hughcars commented on fem/pfespace.cpp:
```diff
@@ -1285,6 +1285,280 @@ void ParFiniteElementSpace::GetExteriorVDofs(Array<int> &ext_dofs,
+      }
+   }
+   // In 2D: No artificial boundary detection needed
+   // Boundary elements ARE edges, so no geometric complexity or artificial boundaries
```
This is not true for collections with vertex DoFs. With `ND_R2D_FECollection`, two selected segments split across ranks each retain their shared vertex, although the serial path removes it as curve-interior. Reconcile occurrence parity across the sharing group.

```suggestion
   else if (dim == 2)
   {
      Array<int> boundary_dof_count(GetVSize());
      boundary_dof_count = 0;
      for (const int dof : boundary_edge_dofs)
      {
         boundary_dof_count[dof] = 1;
      }

      gcomm->Reduce<int>(boundary_dof_count, GroupCommunicator::Sum);
      gcomm->Bcast(boundary_dof_count);

      for (const int dof : boundary_edge_dofs)
      {
         if (boundary_dof_count[dof] % 2 == 0)
         {
            dofs_to_remove.insert(dof);
         }
      }
   }
```

Please add an MPI test that compares the parallel result with the serial result for a closed boundary split between ranks. For example:

```cpp
TEST_CASE("Boundary loop vertex DoFs are partition invariant in 2D",
          "[Parallel][ParMesh][BoundaryEdgeDoFs][ND_R2D]")
{
   const int nranks = Mpi::WorldSize();
   if (nranks < 2) { return; }

   Mesh mesh =
      Mesh::MakeCartesian2D(1, 1, Element::TRIANGLE, true, 1.0, 1.0);
   REQUIRE(mesh.GetNE() == 2);

   ND_R2D_FECollection fec(1, 2);

   FiniteElementSpace serial_fes(&mesh, &fec);
   Array<int> serial_boundary_elements(mesh.GetNBE());
   for (int i = 0; i < serial_boundary_elements.Size(); i++)
   {
      serial_boundary_elements[i] = i;
   }

   std::unordered_set<int> serial_boundary_dofs;
   std::unordered_map<int, int> serial_dof_to_edge;
   std::unordered_map<int, int> serial_dof_to_boundary;
   std::unordered_map<int, int> serial_dof_to_orientation;

   serial_fes.GetBoundaryLoopEdgeDofs(
      serial_boundary_elements, serial_boundary_dofs,
      serial_dof_to_edge, serial_dof_to_boundary,
      serial_dof_to_orientation);

   int partitioning[2] = {0, 1};
   ParMesh pmesh(MPI_COMM_WORLD, mesh, partitioning);
   ParFiniteElementSpace parallel_fes(&pmesh, &fec);

   Array<int> local_boundary_elements(pmesh.GetNBE());
   for (int i = 0; i < local_boundary_elements.Size(); i++)
   {
      local_boundary_elements[i] = i;
   }

   Array<int> ess_tdofs;
   Array<int> ldof_marker;
   std::unordered_set<int> local_boundary_dofs;

   parallel_fes.GetBoundaryLoopEdgeDofs(
      local_boundary_elements, ess_tdofs, ldof_marker,
      local_boundary_dofs);

   int local_tdof_count = ess_tdofs.Size();
   int global_tdof_count = 0;
   MPI_Allreduce(&local_tdof_count, &global_tdof_count, 1, MPI_INT,
                 MPI_SUM, MPI_COMM_WORLD);

   REQUIRE(global_tdof_count ==
           static_cast<int>(serial_boundary_dofs.size()));
}
```

dnpham23 replied:
You're right. Fixed in [dae8a18]([dae8a18](https://github.com/mfem/mfem/pull/4983/commits/dae8a18c323d4adab99c150a0c18ff454a93f01d)) with the reduce-sum-then-parity approach you
suggested (pfespace.cpp:1454).

The reasoning I put in the comment, to make the invariant explicit: membership
in `boundary_edge_dofs` is the local occurrence parity, since the serial code
erases on the second occurrence. Summing those across the sharing group and
testing `% 2 == 0` therefore recovers the global parity, and even parity means
curve-interior.

For the test I adapted rather than copied your snippet: two issues with it as
written: the `GetBoundaryLoopEdgeDofs` signature doesn't match the current API,
and a 1x1 two-triangle mesh doesn't actually produce a boundary loop split
across ranks. `BoundaryEdgeDoFs2DLoopVertexDoFsPartitionInvariant`
(test_bdr_edgedof.cpp:576) uses a 2x2 quad mesh with block and round-robin
partitionings, comparing the summed parallel true-DoF count against the serial
result. Without the fix it reports 10 vs 8 (the two extra shared vertex DoFs);
with it, it passes on 2, 3, and 4 ranks.

tzanio commented:
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

hughcars commented on general/communication.cpp:
```diff
@@ -1108,6 +1108,116 @@ void GroupCommunicator::ReduceEnd(T *ldata, int layout,
+                  // Create a temporary OpData with just this one DoF
+                  OpData<T> single_opd;
+                  single_opd.ldata = ldata;
+                  single_opd.buf = opd.buf + i;
```
`nldofs` is changed to 1, but `buf` still uses the original neighbor-major stride. `Op` therefore reads `buf[i + j]` rather than `buf[j * N + i]`, mixing different DoFs. Gather the selected strided values before applying the reduction.

```suggestion
                  Array<T> single_buf(opd.nb);
                  for (int j = 0; j < opd.nb; j++)
                  {
                     single_buf[j] = opd.buf[j * opd.nldofs + i];
                  }
                  single_opd.buf = single_buf.GetData();
```

Please add a direct `byGroup` regression with multiple ranks and multiple DoFs in the same group. For example:

```cpp
TEST_CASE("GroupCommunicator ReduceMarked byGroup stride",
          "[Parallel][GroupCommunicator]")
{
   const int rank = Mpi::WorldRank();
   const int nranks = Mpi::WorldSize();
   if (nranks < 3) { return; }

   ListOfIntegerSets groups;

   IntegerSet local_group(1);
   local_group[0] = rank;
   groups.Insert(local_group);

   IntegerSet shared_group(nranks);
   for (int r = 0; r < nranks; r++)
   {
      shared_group[r] = r;
   }
   groups.Insert(shared_group);

   GroupTopology topology(MPI_COMM_WORLD);
   topology.Create(groups, 4983);

   GroupCommunicator comm(topology, GroupCommunicator::byGroup);
   Array<int> ldof_group(2);
   ldof_group = 1;
   comm.Create(ldof_group);

   Array<real_t> values(2);
   values[0] = 10.0 * rank + 1.0;
   values[1] = 100.0 * rank + 2.0;

   Array<int> marker(2);
   marker = 1;

   comm.Reduce<real_t>(values.GetData(), marker,
                       GroupCommunicator::Sum<real_t>);
   comm.Bcast(values);

   const real_t rank_sum = nranks * (nranks - 1) / 2.0;
   REQUIRE(values[0] == MFEM_Approx(10.0 * rank_sum + nranks));
   REQUIRE(values[1] == MFEM_Approx(100.0 * rank_sum + 2.0 * nranks));
}
```

dnpham23 replied:
Confirmed and fixed in [dae8a18](https://github.com/mfem/mfem/pull/4983/commits/dae8a18c323d4adab99c150a0c18ff454a93f01d), using the gather approach you suggested.

For the record on why only this branch was affected: `byNeighbor` builds its
`OpData` with `nb = 1` and a per-group contiguous `buf`, so setting
`nldofs = 1` there is harmless. Only `byGroup` carries `nb = GetGroupSize-1`
with the neighbor-major stride, so it was the sole broken path.

Added `GroupCommunicatorReduceMarkedByGroupStride` (test_bdr_edgedof.cpp:662),
which follows your sketch: 3+ ranks, two DoFs in one shared group. It fails
without the fix and passes with it.

hughcars commented on fem/pfespace.cpp:
```diff
@@ -1285,6 +1285,280 @@ void ParFiniteElementSpace::GetExteriorVDofs(Array<int> &ext_dofs,
+   boundary_edge_dofs_out = boundary_edge_dofs;
+
+   // Convert to true DoFs and output
+   ess_tdof_list.SetSize(0);
```
`ess_tdof_list` is reset on every call, but `ess_edge_list` is only appended to. Reusing the output array leaves stale entries and breaks its correspondence with the true-DoF list.

```suggestion
   ess_tdof_list.SetSize(0);
   if (ess_edge_list)
   {
      ess_edge_list->SetSize(0);
   }
```

dnpham23 replied:
Fixed in [dae8a18](https://github.com/mfem/mfem/pull/4983/commits/dae8a18c323d4adab99c150a0c18ff454a93f01d) (pfespace.cpp:1505).
Verified the reuse behavior directly: calling the method three times with the
same output arrays previously grew `ess_edge_list` 8 -> 16 -> 24 while
`ess_tdof_list` stayed at 8. After the fix both stay at 8 across repeated calls.

dnpham23 committed:
Address review feedback

- communication.hpp: discard the BitOR_Op branch at compile time for
  non-integral types, which otherwise failed to link.
- communication.cpp: fix the neighbor-major buffer stride in
  ReduceMarked() byGroup mode, which mixed values from different DoFs.
- pfespace.cpp: reset ess_edge_list alongside ess_tdof_list so reuse
  does not leave stale entries.
- pfespace.cpp: reconcile 2D boundary vertex DoF occurrence parity
  across sharing groups, so a loop split between ranks matches serial.
- pfespace.cpp: synchronize the local marker before converting to true
  DoFs, so a selected shared DoF is not skipped by every rank.

Adds three regression tests, each confirmed to fail without its fix.

dylan-copeland commented on tests/unit/fem/test_bdr_edgedof.cpp:
```diff
@@ -0,0 +1,713 @@
+
+#ifdef MFEM_USE_MPI
+
+namespace boundary_edge_dof_test
```
This namespace and `GeneratePartitionings` are unused and can be removed now, since the partition choice was changed.

dnpham23 replied:
Removed. This was leftover from an earlier exhaustive-partitioning approach and is no longer used.

dylan-copeland commented on tests/unit/fem/test_bdr_edgedof.cpp:
```diff
@@ -0,0 +1,713 @@
+      {
+         block[i] = (i < n_elements/2) ? 0 : test_num_procs-1;
+      }
+      all_partitionings.push_back(block);
```
It shouldn't matter much for a small test, but `push_back` makes a copy of `block`. To avoid duplicating data, `block` could be a reference to `all_partitionings[1]` after setting the size of `all_partitionings` accordingly.

dnpham23 replied:
Good catch. Switched to `all_partitionings.emplace_back(n_elements)` and filling the returned reference in place, so no copy.

dylan-copeland commented on tests/unit/fem/test_bdr_edgedof.cpp:
```diff
@@ -0,0 +1,713 @@
+      all_results.push_back(total_dofs);
+   }
+
+   // Verify all results are identical
```
Please add a comment explaining why matching the total number of DOFs means the test was successful.

dnpham23 replied:
Added a comment: the boundary edge DOF set is a property of the mesh geometry, independent of element distribution, so a correct implementation yields the same global count for every partitioning; a mismatch means some partition kept/dropped a DOF another did not.

dylan-copeland commented on tests/unit/fem/test_bdr_edgedof.cpp:
```diff
@@ -0,0 +1,713 @@
+            return std::set<HYPRE_BigInt>(all.begin(), all.end());
+         };
+
+         const std::set<HYPRE_BigInt> global_selected = all_gather(selected);
```
A high-level comment on what is being checked here would be helpful.

dnpham23 replied:
I added a comment explaining that `global_selected` is every shared boundary DOF chosen on any rank and `global_emitted` is every true DOF actually placed in some rank's `ess_tdof_list`; a selected DOF missing from `global_emitted` is one no rank owns and outputs — exactly the synchronization bug this test guards against.

dylan-copeland commented on general/communication.hpp:
```diff
@@ -416,6 +417,54 @@ class GroupCommunicator
+   }
+
+   // Enum for reduction operations
+   enum ReduceOp { Sum_Op, Min_Op, Max_Op, BitOR_Op, MaxAbs_Op };
```
Can this be changed to `enum class`?

dnpham23 replied:
Done! changed to enum class ReduceOp and qualified the use sites.

dylan-copeland commented on fem/fespace.hpp:
```diff
@@ -1389,6 +1390,59 @@ class FiniteElementSpace
+   /** @brief Extract edge degrees of freedom from specified boundary elements.
+       This method identifies all edge-based degrees of freedom that lie on the
+       boundary edges of the specified boundary elements.
+       @param[in] boundary_element_indices Array of all boundary element indices on a certain
```
There is an informal maximum of 80 columns per line. These lines exceed 80.

dnpham23 replied:
Reflowed the doc comment to stay within 80 columns.

dylan-copeland commented on fem/fespace.hpp:
```diff
@@ -1389,6 +1390,59 @@ class FiniteElementSpace
+       boundary edges of the specified boundary elements.
+       @param[in] boundary_element_indices Array of all boundary element indices on a certain
+       boundary surface (3D) or boundary curve (2D)
+       @param[out] boundary_edge_dofs Set of local DoF indices that lie on boundary edges
```
MFEM already has mixed usage of `DOF` and `DoF`, but a `grep` in `mfem/fem` shows `DOF` is much more common. Changing to `DOF` throughout this PR would be more consistent.

dnpham23 replied:
I renamed `DoF` -> `DOF` throughout this PR's additions (comments, test-case names, tags).

dylan-copeland commented on fem/fespace.hpp:
```diff
@@ -1389,6 +1390,59 @@ class FiniteElementSpace
+       @note Uses a toggle approach: DoFs appearing on multiple boundary edges are considered
+       interior to the boundary surface/curve and are excluded from the result.
+       @note This method uses GetEdgeDofs internally, which returns both vertex and edge DoFs.
+       For standard Nédélec elements (ND_FECollection), vertex DoFs are zero, so only actual
```
`vertex DoFs are zero` -> `there are no vertex DOFs`?

dnpham23 replied:
Reworded to "there are no vertex DOFs" for `ND_FECollection`.

dylan-copeland commented on fem/fespace.cpp:
```diff
@@ -4527,6 +4529,230 @@ ::GetEssentialBdrEdgesFaces(const Array<int> &bdr_attr_is_ess,
+               }
+               else
+               {
+                  // DoF appears twice - interior to boundary, remove it
```
My understanding of this logic is that in 3D, an interior boundary surface should have boundary edges identified as being contained by only 1 boundary face. By inserting to and erasing from `boundary_edge_dofs`, the code is checking whether the number of containing faces per edge is odd, not necessarily 1. What if an interior boundary surface is T-shaped, so an interior edge to the surface has 3 incident faces? Should the check be for 1 containing face rather than an odd number, for more generality? It looks like `pfespace.cpp` already does this, so the serial implementation should do so as well.

dnpham23 replied:
Thank you for pointing it out. The parity toggle was inconsistent with the parallel path and would misclassify a non-manifold junction where an edge is shared by three faces. I rewrote both the 3D and 2D serial branches to count how many selected boundary elements contain each DOF and keep only the DOFs seen exactly once, matching the parallel "edge in >= 2 faces -> remove" rule.

dylan-copeland commented on fem/fespace.cpp:
```diff
@@ -4527,6 +4529,230 @@ ::GetEssentialBdrEdgesFaces(const Array<int> &bdr_attr_is_ess,
+         for (int k = 0; k < edge_dofs.Size(); ++k)
+         {
+            const int dof = edge_dofs[k];
+            if (!boundary_edge_dofs.count(dof))
```
In 3D, I see that counting the number of boundary faces containing an edge indicates whether the edge is interior to the boundary surface. What is the logic here, in 2D? In 2D, shouldn't every boundary edge occur exactly once on the boundary? Or do interior boundary edges occur twice?

dnpham23 replied:
I added a comment. In 2D each boundary element is a single edge, so a genuine interior edge DOF belongs to exactly one segment. Adjacent segments share only their common vertex, so a vertex DOF (present for collections like ND_R2D) is seen twice (once per segment) and is correctly dropped by the same count == 1 rule.

dylan-copeland replied:
Thanks for explaining more, but it still seems to me that all interior edge DOFs would be found here, in all cases? Could you just use `GetEdgeInteriorDofs` instead of counting occurrences, or is this different somehow?

dnpham23 replied:
Thanks for giving this a lot of thoughts! You're right that for the edge-interior DOFs the occurrence count is really just identifying which edges lie on the loop boundary; each edge-interior DOF belongs to a single edge, and a loop-boundary edge belongs to exactly one selected face (3D) / segment (2D), so those DOFs always end up with count == 1.

The reason I can't switch to `GetEdgeInteriorDofs` is the vertex DOFs. `GetEdgeDofs` returns the two endpoint vertex DOFs in addition to the edge-interior ones, whereas `GetEdgeInteriorDofs` returns only the edge-interior DOFs. For collections that carry vertex DOFs (e.g. ND_R2D), a vertex shared by two loop elements is interior to the boundary loop and must be dropped, while a genuine loop-endpoint vertex must be kept. That distinction is what the count == 1 vs count > 1 test captures (the shared vertex is seen twice, the endpoint once). If I used `GetEdgeInteriorDofs`, all vertex DOFs would be excluded, including the loop endpoints the method documents as kept, and the ND_R2D case that `BoundaryEdgeDOFs2DLoopVertexDOFsPartitionInvariant` covers would regress.

I confirmed this on a 2-by-2 quad mesh with ND_R2D: an open curve keeps its 2 endpoint vertex DOFs (4 DOFs total), a closed loop drops all vertex DOFs (8 DOFs total), and `GetEdgeInteriorDofs` returns no vertex DOFs at all, so it couldn't produce the endpoint vertex DOFs. I've added a note to the comment making this "double duty" of the counting explicit.

dylan-copeland commented on fem/fespace.cpp:
```diff
@@ -4527,6 +4529,230 @@ ::GetEssentialBdrEdgesFaces(const Array<int> &bdr_attr_is_ess,
+   const Array<int> &bdr_attrs,
+   std::unordered_map<int, Array<int>> &attr_to_elements)
+{
+   GetBoundaryElementsByAttributeImpl(mesh, bdr_attrs, attr_to_elements);
```
Do we need `GetBoundaryElementsByAttributeImpl`? Can we just put the code in this function? Same question for `ComputeLoopEdgeOrientationsImpl`. It looks like these functions are also called by `ParFiniteElementSpace`, but that is a derived class and can just use the base class functions here.

dnpham23 replied:
Removed both `GetBoundaryElementsByAttributeImpl` and `ComputeLoopEdgeOrientationsImpl` and inlined their bodies into the member functions in both `fespace.cpp` and `pfespace.cpp`. (They also had external linkage with identical names across the two translation units, so this removes a latent ODR hazard.)

dylan-copeland commented on fem/pfespace.hpp:
```diff
@@ -460,6 +462,57 @@ class ParFiniteElementSpace : public FiniteElementSpace
    void GetExteriorTrueDofs(Array<int> &ext_tdof_list,
                             int component = -1) const override;
 
+   /** @brief Get a list of edge degrees of freedom on the boundary with the specified
```
80-column max.

dnpham23 replied:
Adjusted the affected doc comment lines to 80 columns.

dylan-copeland commented on fem/fespace.hpp:
```diff
@@ -1389,6 +1390,59 @@ class FiniteElementSpace
+
+       @note In 3D: Extracts edge DoFs from the 1D edges of 2D boundary elements (faces).
+       @note In 2D: Extracts edge DoFs directly from 1D boundary elements (segments).
+       In 2D, dof_to_edge and dof_to_boundary_element contain identical edge indices.
```
Is there a way to avoid this duplication of data in these two maps in 2D? Could this be done conveniently by defining a `struct` for the output arguments, in which only one of these maps would contain data in the 2D case?

dnpham23 replied:
Thanks! I can see the appeal of grouping the outputs into one struct. My hesitation is that the specific redundancy you pointed out (`dof_to_edge` and `dof_to_boundary_element` being identical in 2D) comes from the underlying data rather than the parameter list, so a struct would tidy the signature but wouldn't remove the duplicate values themselves. It would also touch both the serial and parallel signatures and all the call sites/tests, which felt like a lot of churn. I'm happy to go ahead with it if you think it's worth it, though. Just let me know and I'll put it together.

dylan-copeland replied:
I was wondering whether the duplicated values could be avoided in the 2D case, which means one of the data structures would be unset, and that this could be handled in a struct. It may be a little complicated to implement something like this just to avoid the data duplication, and the 1D data in 2D shouldn't be too large, so it's not clear this is really needed. It was just a thought, and I'm fine with leaving it as is unless you find a good way to avoid the duplication.

dnpham23 replied:
Thanks for the flexibility. I gave the "unset one structure in 2D" idea some thought, but making one map conditionally empty would push the special-casing onto every caller (they'd have to know which map is valid in 2D vs 3D), which felt like a worse trade than the small amount of duplicated 1D data. I'll leave it as is for now, but I'm open to revisiting if a cleaner approach turns up.

dnpham23 committed:
Address review feedback

- Serial GetBoundaryLoopEdgeDofs (3D and 2D): keep an edge/vertex DOF only
  when it appears in exactly one selected boundary element, matching the
  parallel version and handling non-manifold junctions correctly.
- Inline the GetBoundaryElementsByAttributeImpl and
  ComputeLoopEdgeOrientationsImpl free helpers into their member functions.
- Make GroupCommunicator::ReduceOp an enum class and note that the reduction
  result is signed.
- Rename DoF -> DOF throughout the PR's additions and reflow doc comments to
  80 columns.
- Remove the unused GeneratePartitionings helper; build test partitions in
  place; add explanatory comments to the partition-invariant and
  shared-DOF-ownership tests.

dylan-copeland commented on fem/fespace.cpp:
```diff
@@ -4527,6 +4529,230 @@ ::GetEssentialBdrEdgesFaces(const Array<int> &bdr_attr_is_ess,
+      {
+         const int boundary_element_idx = boundary_element_indices[i];
+         int face_index, face_orientation;
+         mesh->GetBdrElementFace(boundary_element_idx, &face_index, &face_orientation);
```
Sorry we don't have a tool to automate this, but here is another long line.

dnpham23 replied:
Fixed. wrapped the `GetBdrElementFace` call.

dnpham23 committed:
Address review feedback

- Mention float in the Sum/Min/Max/MaxAbs instantiation comments.
- Wrap a long GetBdrElementFace call to stay within 80 columns.
- Explain in GetBoundaryLoopEdgeDofs why occurrences of GetEdgeDofs are
  counted rather than collecting GetEdgeInteriorDofs: the count also resolves
  vertex DOFs (kept at loop endpoints, dropped when shared), which
  GetEdgeInteriorDofs omits entirely.

dylan-copeland reviewed (approved):
Thanks @dnpham23 for this contribution! One small comment that may have been forgotten is that `MaxAbs` in `communication.hpp` could have additional documentation explaining that it returns signed values with maximum absolute value. Otherwise, one might think it returns the non-negative absolute values.

dnpham23 committed:
Address review feedback

Document that GroupCommunicator::MaxAbs returns the signed value with the
largest absolute value, not the non-negative absolute value.

dnpham23 commented:
Thanks for the review and approval, @dylan-copeland! Good catch on MaxAbs. I've updated the doc comment to make that explicit.

dnpham23 committed:
Merge remote-tracking branch 'upstream/master'

dnpham23 commented:
Ping @tzanio as I was wondering if this PR good to merge? It's needed in Palace for magnetostatic solve. Thank you!

najlkin commented on general/communication.hpp:
```diff
@@ -408,14 +409,67 @@ class GroupCommunicator
+         default:                  Op = GroupCommunicator::Sum; break;
+      }
+      Reduce(ldata.GetData(), marker, Op);
+   }
```
What is the motivation for this enum version? The only benefit I see here is the check on bitor, which can be done in `BitOR` template directly. I would remove this version and keep only the other one.

dnpham23 replied:
I removed the enum overload; the only useful part was the integral check, which is now a `static_assert` directly inside the `BitOR` template.

najlkin commented on fem/pfespace.hpp:
```diff
@@ -460,6 +462,61 @@ class ParFiniteElementSpace : public FiniteElementSpace
+       artificial boundary edges that appear at processor boundaries.
+       Requirements:
+       - Mesh must be conforming (no hanging nodes)
+       - Mesh dimension must be >= 2
```
The concept of the "loop" is not explained here.

dnpham23 replied:
I added a technical definition of "loop" to the docs: the boundary edges bounding the region covered by the selected boundary elements; an edge shared by two selected elements is interior and excluded. (Keeping the exclusion behavior as the defining feature rather than an option, per the documented semantics.)

najlkin commented on fem/pfespace.hpp:
```diff
@@ -460,6 +462,61 @@ class ParFiniteElementSpace : public FiniteElementSpace
+   void GetBoundaryLoopEdgeDofs(const Array<int> &boundary_element_indices,
+                                Array<int> &ess_tdof_list,
+                                Array<int> &ldof_marker,
+                                std::unordered_set<int> &boundary_edge_dofs_out,
```
All these are essentially carrying the same information and the user probably needs just one variant of it. We have usually VDofs and TrueDofs versions of methods, so maybe something like that would work with a single internal implementation. Alternatively, these parameters could be made optional 🤔 .

dnpham23 replied:
Thanks for a good point! I consolidated to a single ordering with optional `Array<int>` out-params, so callers take only what they need.

najlkin replied:
I do not see it, the user needs one most likely, but will get three, where especially `ldof_marker` is a pretty big array. See `ListToMarker()` and `MarkerToList()`.

dnpham23 replied:
Done. `ldof_marker` is now an optional trailing` Array<int>*` output (default `nullptr`), so callers who only want `ess_tdof_list`/`boundary_edge_dofs` no longer pay for the `GetVSize()`-sized array. It's still computed internally where the reconciliation needs it and only swapped into the caller's array on request. I noted in the docs that it's derivable from `boundary_edge_dofs` via `ListToMarker()`.

najlkin commented on fem/fespace.hpp:
```diff
@@ -1389,6 +1390,66 @@ class FiniteElementSpace
+                                std::unordered_set<int> &boundary_edge_dofs,
+                                std::unordered_map<int, int> &dof_to_edge,
+                                std::unordered_map<int, int> &dof_to_boundary_element,
+                                std::unordered_map<int, int> &dof_to_edge_orientation) const;
```
These are not much typical MFEM objects and the maps have unnecessary overhead with hashing etc. I would consider just giving the DOFs an arbitrary order (perhaps how you add them to the set) and have just `Array<int>` for them. The rest can be arrays as well, following the same ordering/indexing.

dnpham23 replied:
Done! The interface now exposes only `Array<int>` outputs sharing a single ordering (sorted by DOF index for determinism); the `std::unordered_map`/`set` are kept only as internal scratch in the `.cpp`. `dof_edges`/`dof_boundary_elements` follow the same indexing as the DOF array.

najlkin commented on fem/fespace.hpp:
```diff
@@ -1389,6 +1390,66 @@ class FiniteElementSpace
+       Serial version that finds all boundary elements with specified attributes. */
+   virtual void GetBoundaryElementsByAttribute(
+      const Array<int> &bdr_attrs,
+      std::unordered_map<int, Array<int>> &attr_to_elements);
```
The same as above ☝️ .

dnpham23 replied:
Fixed above ^

najlkin replied:
I mean it can be just `std::vector<Array<int>>` or something like that, the index can be the same as in the array `bdr_attrs`.

dnpham23 replied:
Done. Changed the output to `std::vector<Array<int>>`, indexed the same as the input `bdr_attrs` (`attr_to_elements[i] `corresponds to `bdr_attrs[i]`), as suggested.

najlkin commented on fem/fespace.hpp:
```diff
@@ -1389,6 +1390,66 @@ class FiniteElementSpace
+      const std::unordered_map<int, int>& dof_to_boundary_element,
+      const Vector& loop_normal,
+      std::unordered_map<int, int>& edge_loop_orientations);
+
```
Here too. Besides, this would need description of the parameters and reference to `GetBoundaryLoopEdgeDofs()`.

dnpham23 replied:
Fixed! See the reply above.

najlkin commented on fem/fespace.hpp:
```diff
@@ -1389,6 +1390,66 @@ class FiniteElementSpace
+       (segments). In 2D, dof_to_edge and dof_to_boundary_element contain
+       identical edge indices.
+       @note A DOF that lies in more than one of the selected boundary elements is
+       interior to the boundary surface/curve and is excluded from the result.
```
This makes sense in your application, but it is somewhat specific, I would consider making this behavior optional and the name of the method more general. Alternatively, this should be explained as the technical definition of the "loop".

dnpham23 replied:
A technical definition of "loop" is added to the docs.

dnpham23 replied:
Done! `dof_to_orientation` is gone. Orientation is computed on demand by `ComputeLoopEdgeOrientations`, which returns a parallel-indexed Array<int>.

najlkin reviewed (commented):
A good amount of work done here! 👍 I was not checking much the implementation, mostly the overall interface and I have some comments below:

tzanio commented:
ping: @dnpham23, @hughcars

tzanio commented:
Useful feature, but at a minimum it should have a `CHANGELOG` entry and more documentation.

tzanio commented on tests/unit/fem/test_bdr_edgedof.cpp:
```diff
@@ -0,0 +1,692 @@
+                                    real_t(2.0) * real_t(nranks)));
+}
+
+#endif // MFEM_USE_MPI
```
```suggestion
#endif // MFEM_USE_MPI

```

dnpham23 replied:
Already present. Thanks!

v-dobrev replied:
The suggestion adds a missing newline at the end of the file.

dnpham23 replied:
Ah sorry i missed it. newline added.

tzanio commented on tests/unit/fem/test_bdr_edgedof.cpp:
```diff
@@ -0,0 +1,692 @@
+   // processor boundaries, every partitioning yields the same total. A mismatch
+   // means some partition kept or dropped a DOF that another did not.
+   REQUIRE(!all_results.empty());
+   int expected = all_results[0];
```
isn't it better to just use 6 here?

dnpham23 replied:
Yes, done!

tzanio commented on fem/pfespace.hpp:
```diff
@@ -460,6 +462,61 @@ class ParFiniteElementSpace : public FiniteElementSpace
+                                    dof_to_edge,
+                                    const std::unordered_map<int, int>& dof_to_boundary_element,
+                                    const Vector& loop_normal,
+                                    std::unordered_map<int, int>& edge_loop_orientations) override;
```
Why do we need these in `ParFiniteElementSpace`? They don't actually change anything compared to the serial versions, can't we simply inherit the base implementations?

dnpham23 replied:
Great observation, thank you! All Par constructors forward the same `ParMesh*` to both the base ctor and `ParInit`, so `mesh == pmesh` and the base implementations already operate on the `ParMesh`. The base versions are now inherited.

tzanio commented on general/communication.hpp:
```diff
@@ -408,14 +409,67 @@ class GroupCommunicator
+   /// Reduce operation selecting the signed value with the largest absolute
+   /// value, instantiated for int, double and float. Note the result keeps its
+   /// sign; it is not the non-negative absolute value.
+   template <class T> static void MaxAbs(OpData<T>);
```
This is a strange operation and a bit brittle: what happens if there are two max entries with opposite signs?

dnpham23 replied:
Documented: `MaxAbs` keeps the signed value of largest magnitude, and ties are broken deterministically by a strict "greater magnitude" test, so equal-magnitude contributions (including opposite signs) keep the value already accumulated at the master.

najlkin replied:
Yeah, I agree this is a strange operation, which depends on the ordering and which is not used in this PR anymore except the unit test (some discussion was here https://github.com/mfem/mfem/pull/4983#discussion_r2350515035). It is an independent feature, which should be mentioned separately in Changelog. However, I am more leaning to removing it all together. What do you think @tzanio , @dylan-copeland and @mlstowell ?

dnpham23 replied:
Thanks! I'm happy to give `MaxAbs` its own Changelog entry and document it as a standalone, general-purpose `GroupCommunicator` reduction rather than something tied to the (removed) `SynchronizeBC`.

It's needed kinda imminently downstream: in Palace's magnetostatic solver we impose magnetic-flux boundary conditions on boundaries of metallic holes, where signed boundary values set on `ldofs` must be inherited by the shared `tdofs` preserving both magnitude and sign. `MaxAbs` does that in a single reduction; the workaround is an array copy plus two reductions (`Max` and `Min`) compared afterward, which is more code on the communication path for the same result.

On the ordering/brittleness point: the only order-sensitive case is an exact equal-magnitude tie with opposite signs (|+a| == |-a|); every other input gives a unique result, and even that tie is deterministic for a fixed mesh partition (group ownership and buffer order are fixed). It's no more order-dependent than the existing floating-point `Sum`/`Max` reductions. In the BC use case all contributions to a shared `tdof` carry the same physical value and sign, so opposite-sign ties don't arise. If you'd prefer a fully partition-invariant rule regardless, I'm happy to break ties deterministically toward the positive value (a one-line change).

My preference is to keep it here with its own changelog line since it's already reviewed, but if you'd all rather it land separately I can split it into a small standalone PR. Whatever you prefer.

tzanio replied:
@dnpham23, is it possible to change it so if there are two max entries with opposite signs, we return the positive one? I just think that it will be better to have a clear deterministic outcome.

najlkin replied:
At least something like would be sufficient imho + Changelog. I see it is useful for checking validity of the BC projection, even when a proper procedure should not rely on that, like the fixed tangential projection in #5280 .

dnpham23 replied:
Thank you! `MaxAbs` now breaks equal-magnitude ties deterministically toward the positive value. Implemented as a one-line change + doc comment updated. I also added a unit test (`GroupCommunicatorMaxAbs`) that pins both behaviors: an opposite-sign tie held by the group master resolves to the positive value, and a larger-magnitude negative held by a peer wins and retains its sign. Changelog entry is added as well.

najlkin commented on fem/fespace.hpp:
```diff
@@ -22,6 +22,7 @@
 #include "restriction.hpp"
 #include <iostream>
 #include <unordered_map>
+#include <unordered_set>
```
```suggestion
```

dnpham23 replied:
Done. Removed the unused `<unordered_set>` from both headers. I also removed a couple of related redundant includes while I was in there: `<unordered_map>` was unused in `pfespace.hpp`, and `<unordered_set>` was unused in `fespace.cpp`. `pfespace.cpp` now includes `<unordered_map>`/`<unordered_set>` directly (it uses both) rather than picking them up transitively. `fespace.hpp` keeps `<unordered_map>` (used by the face-restriction/interpolation members) and now includes `<vector>` for the changed API below.

najlkin commented on fem/pfespace.hpp:
```diff
@@ -20,6 +20,8 @@
 #include "../mesh/nurbs.hpp"
 #include "fespace.hpp"
+#include <unordered_map>
+#include <unordered_set>
```
```suggestion
```

dnpham23 replied:
Done. Please see comment for `fespace.hpp` above.

tzanio commented:
@dnpham23, since I can't push to your branch can you please make these final changes before we merge in `next` for testing?

1️⃣  Remove whitespace

```diff
diff --git a/data/square_in_square.geo b/data/square_in_square.geo
--- a/data/square_in_square.geo
+++ b/data/square_in_square.geo
@@ -60,7 +60,7 @@ Physical Surface(2) = {2}; // Inner material (solid square)
 // Physical lines for boundary conditions
 // Outer square boundary edges
 Physical Line(1) = {1}; // outer bottom
-Physical Line(2) = {2}; // outer right  
+Physical Line(2) = {2}; // outer right
 Physical Line(3) = {3}; // outer top
 Physical Line(4) = {4}; // outer left
```

2️⃣ Count H1 dofs shared by multiple edges only once
```diff
diff --git a/fem/fespace.cpp b/fem/fespace.cpp
--- a/fem/fespace.cpp
+++ b/fem/fespace.cpp
@@ -4563,6 +4563,7 @@ void FiniteElementSpace::GetBoundaryLoopEdgeDofs(
    for (int i = 0; i < boundary_element_indices.Size(); ++i)
    {
       const int boundary_element_idx = boundary_element_indices[i];
+      std::unordered_set<int> boundary_element_dofs;
 
       if (dim == 3)
       {
@@ -4587,7 +4588,8 @@ void FiniteElementSpace::GetBoundaryLoopEdgeDofs(
          {
             const int dof = edge_dofs[k];
             // Record metadata the first time this DOF is seen.
-            if (dof_count[dof]++ == 0)
+            if (boundary_element_dofs.insert(dof).second &&
+                dof_count[dof]++ == 0)
             {
                dof_edge[dof] = edges[j];
                dof_belem[dof] = boundary_element_idx;
diff --git a/fem/fespace.hpp b/fem/fespace.hpp
--- a/fem/fespace.hpp
+++ b/fem/fespace.hpp
@@ -22,6 +22,7 @@
 #include "restriction.hpp"
 #include <iostream>
 #include <unordered_map>
+#include <unordered_set>
 #include <vector>
 
 namespace mfem
```

3️⃣  Fix for single precision builds
```diff
diff --git a/tests/unit/fem/test_bdr_edgedof.cpp b/tests/unit/fem/test_bdr_edgedof.cpp
index 9c64177851..27a039cc4f 100644
--- a/tests/unit/fem/test_bdr_edgedof.cpp
+++ b/tests/unit/fem/test_bdr_edgedof.cpp
@@ -303,9 +303,11 @@ TEST_CASE("BoundaryEdgeDOFsNestedCubes",
       real_t global_tdof_sum, total_length;
       MPI_Allreduce(&local_nonzero_tdofs, &global_nonzero_tdofs, 1, MPI_INT, MPI_SUM,
                     MPI_COMM_WORLD);
-      MPI_Allreduce(&local_tdof_sum, &global_tdof_sum, 1, MPI_DOUBLE, MPI_SUM,
+      MPI_Allreduce(&local_tdof_sum, &global_tdof_sum, 1,
+                    MPITypeMap<real_t>::mpi_type, MPI_SUM,
                     MPI_COMM_WORLD);
-      MPI_Allreduce(&local_length, &total_length, 1, MPI_DOUBLE, MPI_SUM,
+      MPI_Allreduce(&local_length, &total_length, 1,
+                    MPITypeMap<real_t>::mpi_type, MPI_SUM,
                     MPI_COMM_WORLD);
 
       // Verify processor-invariant results match expected values
```

dnpham23 commented:
@tzanio Done, all three in [fcbbfd49b9](https://github.com/mfem/mfem/pull/4983/commits/fcbbfd49b901538d6230cbbceaa5fdf43f7dc1a8).

tzanio commented:
Merged in `next` for testing...
